crates/sui-rest-api/src/transactions/resolve.rs

```diff
@@ -0,0 +1,564 @@
+            .inner()
+            .get_object_by_key(&id, version.into())?
+            .ok_or_else(|| ObjectNotFoundError::new_with_version(object_id, version))?;
+        (object.version(), object.digest())
+    } else {
+        let object = reader
+            .inner()
+            .get_object(&id)?
+            .ok_or_else(|| ObjectNotFoundError::new(object_id))?;
+        (object.version(), object.digest())
+    };
+
+    if digest.is_some_and(|digest| digest.inner() != d.inner()) {
+        return Err(RestError::new(
+            axum::http::StatusCode::BAD_REQUEST,
+            format!("provided digest doesn't match, provided: {digest:?} actual: {d}"),
+        ));
+    }
+
+    Ok((id, v, d))
+}
+
+fn resolve_ptb(
+    reader: &StateReader,
+    called_packages: &HashMap<ObjectId, NormalizedPackage>,
+    unresolved_ptb: UnresolvedProgrammableTransaction,
+) -> Result<ProgrammableTransaction> {
+    let inputs = unresolved_ptb
+        .inputs
+        .into_iter()
+        .enumerate()
+        .map(|(arg_idx, arg)| {
+            resolve_arg(
+                reader,
+                called_packages,
+                &unresolved_ptb.commands,
+                arg,
+                arg_idx,
+            )
+        })
+        .collect::<Result<_>>()?;
+
+    ProgrammableTransaction {
+        inputs,
+        commands: unresolved_ptb
+            .commands
+            .into_iter()
+            .map(Into::into)
+            .collect(),
+    }
+    .pipe(Ok)
+}
+
+fn resolve_arg(
+    reader: &StateReader,
+    called_packages: &HashMap<ObjectId, NormalizedPackage>,
+    commands: &[Command],
+    arg: UnresolvedInputArgument,
+    arg_idx: usize,
+) -> Result<CallArg> {
+    match arg {
+        UnresolvedInputArgument::Pure { value } => CallArg::Pure(value),
+        UnresolvedInputArgument::ImmutableOrOwned(obj_ref) => CallArg::Object(
+            ObjectArg::ImmOrOwnedObject(resolve_object_reference(reader, obj_ref)?),
+        ),
+        UnresolvedInputArgument::Shared {
+            object_id,
+            initial_shared_version: _,
+            mutable: _,
+        } => {
+            let id = object_id.into();
+            let object = reader
+                .inner()
+                .get_object(&id)?
+                .ok_or_else(|| ObjectNotFoundError::new(object_id))?;
+
+            let initial_shared_version = if let sui_types::object::Owner::Shared {
+                initial_shared_version,
+            } = object.owner()
+            {
+                *initial_shared_version
+            } else {
+                return Err(RestError::new(
+                    axum::http::StatusCode::BAD_REQUEST,
+                    format!("object {object_id} is not a shared object"),
+                ));
+            };
+
+            let mut mutable = false;
+
+            for (move_call, idx) in find_arg_uses(arg_idx, commands).filter_map(|(command, idx)| {
+                if let Command::MoveCall(move_call) = command {
```

Comment:
recently learned that other commands (splitCoin, mergeCoins, transferObjects) can also take shared objects and they should be marked as mutable

Reply:
is this true? i'm not sure i understand how that works. Do you have an example of this?

Reply:
Whoops, this is partially inaccurate.  It should have been makeMoveVec instead of transferObjects.

We did have a user report issues with using shared objects in a makeMoveVec.  I don't have examples of the split/merge coins example, but was told it was possible for the the source coin in a split, and the first coin in a merge to be a shared mutable coin.

Reply:
here's a tx with a split and merge of a shared coin: https://suiscan.xyz/devnet/tx/2a1W7uJSkMNc135CH6M5dPc9y6NRmMmVzGVr7VwS8rLt